agent/php_user_instrument.c

```diff
@@ -644,89 +509,71 @@ void nr_php_user_function_add_declared_callback(const char* namestr,
- *
- *  4. When a zend function is called and the agent tries to obtain the
- *     wraprec, the upper bits of the value are compared to the current process
- *     id. If they match, the index in the lower 16 bits is considered safe and
- *     is used. Otherwise the function is considered as uninstrumented.
- */
+void nr_php_user_instrument_set(zend_function* func, nruserfn_t* wraprec) {
 
-void nr_php_op_array_set_wraprec(zend_op_array* op_array,
-                                 nruserfn_t* func TSRMLS_DC) {
-  uintptr_t index;
+#if LOOKUP_METHOD == LOOKUP_USE_OP_ARRAY
 
-  if (NULL == op_array || NULL == func) {
-    return;
-  }
+  nr_php_op_array_set_wraprec(&func->op_array, wraprec);
 
-  if (!nr_vector_push_back(NRPRG(user_function_wrappers), func)) {
-    return;
-  }
+#elif LOOKUP_METHOD == LOOKUP_USE_LINKED_LIST
+
+  wraprec_metadata_set(wraprec, func);
+
+#elif LOOKUP_METHOD == LOOKUP_USE_UTIL_HASHMAP
+
+  util_hashmap_set_wraprec(func, wraprec);
+
+#elif LOOKUP_METHOD == LOOKUP_USE_WRAPREC_HASHMAP
 
-  index = nr_vector_size(NRPRG(user_function_wrappers)) - 1;
+  wraprec_hashmap_set(NRPRG(user_function_wrappers), wraprec, func);
 
-  index |= (NRPRG(pid) << 16);
+#else
+
+#error "Unknown wraprec lookup method"
 
-  op_array->reserved[NR_PHP_PROCESS_GLOBALS(zend_offset)] = (void*)index;
+#endif
 }
 
-#if ZEND_MODULE_API_NO < ZEND_7_4_X_API_NO
-nruserfn_t* nr_php_op_array_get_wraprec(
-    const zend_op_array* op_array TSRMLS_DC) {
-  uintptr_t index;
-  uint64_t pid;
+nruserfn_t* nr_php_user_instrument_get(zend_function* func) {
+  nruserfn_t* p = NULL;
+  unsigned n = 0;
+  struct timespec tstart={0,0}, tend={0,0};
 
-  if (nrunlikely(NULL == op_array)) {
-    return NULL;
+  if (nrunlikely(NR_PHP_PROCESS_GLOBALS(special_flags).debug_user_instrument_lookup)) {
+    clock_gettime(CLOCK_MONOTONIC, &tstart);
   }
 
-  index = (uintptr_t)op_array->reserved[NR_PHP_PROCESS_GLOBALS(zend_offset)];
-
-  if (0 == index) {
+  if ((NULL == func) || (ZEND_USER_FUNCTION != func->type)) {
     return NULL;
   }
 
-  pid = index >> 16;
-  index &= 0xffff;
+#if LOOKUP_METHOD == LOOKUP_USE_OP_ARRAY
 
-  if (pid != NRPRG(pid)) {
-    nrl_verbosedebug(
-        NRL_INSTRUMENT,
-        "Skipping instrumented function: pid mismatch, got " NR_INT64_FMT
-        ", expected " NR_INT64_FMT,
-        pid, NRPRG(pid));
-    return NULL;
-  }
+  n++;
+  p = nr_php_op_array_get_wraprec(&func->op_array);
+
+#elif LOOKUP_METHOD == LOOKUP_USE_LINKED_LIST
+
+  p = nr_php_get_wraprec_by_func(&n, func);
+
+#elif LOOKUP_METHOD == LOOKUP_USE_UTIL_HASHMAP
+
+  n++;
+  p = util_hashmap_get_wraprec(func);
+
+#elif LOOKUP_METHOD == LOOKUP_USE_WRAPREC_HASHMAP
 
-  return (nruserfn_t*)nr_vector_get(NRPRG(user_function_wrappers), index);
+  p = wraprec_hashmap_get(&n, NRPRG(user_function_wrappers), func);
+
+#else
+
+#error "Unknown wraprec lookup method"
+
+#endif
+
+  if (nrunlikely(NR_PHP_PROCESS_GLOBALS(special_flags).debug_user_instrument_lookup)) {
```

Comment:
If this is staying in the PR perhaps the output should go to the agent log file?

Reply:
This probably shouldn't stay in the PR.